From 1d2317beda367ee2d42741da3808197a812ea616 Mon Sep 17 00:00:00 2001
From: pengyuding <pengyuding@chiphd.com>
Date: Thu, 24 Mar 2016 18:38:34 +0800
Subject: [PATCH] move the definition of ro.product.firmware to tulip-chiphd.mk

---
 tulip-common.mk | 2 --
 1 file changed, 2 deletions(-)

diff --git a/tulip-common.mk b/tulip-common.mk
index cdf3a3b..5b4c103 100755
--- a/tulip-common.mk
+++ b/tulip-common.mk
@@ -109,8 +109,6 @@ PRODUCT_PROPERTY_OVERRIDES += \
 # each product.
 PRODUCT_TAGS += dalvik.gc.type-precise
 
-PRODUCT_PROPERTY_OVERRIDES += \
-    ro.product.firmware=v1.2
 
 # if DISPLAY_BUILD_NUMBER := true then
 # BUILD_DISPLAY_ID := $(BUILD_ID).$(BUILD_NUMBER)
-- 
1.9.1

